### ISSUE TITLE
Double free when adding non-existing export

### ISSUE THREAD

**Issue by dirk1730** [nfs-ganesha/nfs-ganesha#258]
Scenario: Add an export to a running ganesha server using a DBUS add command.
The fsal lookup_path function returns `fsalstat(ERR_FSAL_NOENT, ENOENT)` (or ERR_FSAL_SERVERFAULT).
Result: The Ganesha code does a double free. Strangely enough, jemalloc does not crash at that time (using libc does crash).

Call stack for first free:
```
1  free_export              export_mgr.c     249  0x51d207       
2  put_gsh_export           export_mgr.c     640  0x51e704       
3  export_commit_common     exports.c        1344 0x5099f1       
4  add_export_commit        exports.c        1388 0x509bf3       
5  proc_block               config_parsing.c 1336 0x558d2f       
6  load_config_from_node    config_parsing.c 1835 0x55a1ce       
7  gsh_export_addexport     export_mgr.c     975  0x51f517       
8  dbus_message_entrypoint  dbus_server.c    512  0x553957       
9  ??                                             0x7ffff7bac813 
10 dbus_connection_dispatch                       0x7ffff7b9dd94 
11 ??                                             0x7ffff7b9e09d 
12 gsh_dbus_thread          dbus_server.c    738  0x5545e0       
13 start_thread             pthread_create.c 333  0x7ffff70516ba 
14 clone                    clone.S          109  0x7ffff693041d 
```

Call stack for second free:
```
1  free_export              export_mgr.c     249  0x51d207       
2  export_init              exports.c        947  0x5084e8       
3  proc_block               config_parsing.c 1359 0x558df5       
4  load_config_from_node    config_parsing.c 1835 0x55a1ce       
5  gsh_export_addexport     export_mgr.c     975  0x51f517       
6  dbus_message_entrypoint  dbus_server.c    512  0x553957       
7  ??                                             0x7ffff7bac813 
8  dbus_connection_dispatch                       0x7ffff7b9dd94 
9  ??                                             0x7ffff7b9e09d 
10 gsh_dbus_thread          dbus_server.c    738  0x5545e0       
11 start_thread             pthread_create.c 333  0x7ffff70516ba 
12 clone                    clone.S          109  0x7ffff693041d 
```

**Comment by dang** [comment]
The line numbers don't line up with 2.6 or 2.5.  What version of Ganesha are you using?

**Comment by dirk1730** [comment]
According to the ChangeLog: 2.3.0.

**Comment by dirk1730** [comment]
I just noticed that the changelog didn't get updated anymore since 27 Oct 2015.
I believe I downloaded the files somewhere around June 2017. Not sure which version that is.

**Comment by supriti** [comment]
if you run /usr/bin/ganesha.nfsd -v that should give you the version number.

**Comment by dirk1730** [comment]
2.5.0.2

**Comment by dirk1730** [comment]
Some more info about the stack traces:

config_parsing.c:1835
`if (!proc_block(node, &conf_blk->blk_desc, param, err_type)) {`

config_parsing.c:1336
`errors = item->u.blk.commit(node, link_mem, param_struct, err_type);`

config_parsing.c:1359
```
err_out:
(void)item->u.blk.init(link_mem, param_struct);
return false;
```

exports.c:947
`free_export(export);`

exports.c:1344
`put_gsh_export(export);`

exports.c:1388
`return export_commit_common(node, link_mem, self_struct, err_type, add_export);`

export_mgr.c:249
`assert(export->refcnt == 0);`

export_mgr.c:640
`free_export(export);`

Hope this helps.

Dirk.

**Comment by dang** [comment]
Update to the newest version of 2.5.  This was fixed by 9d741775c71d1ff52e7444f3f1d791301fa58b27

**Comment by dirk1730** [comment]
Thanks a lot! We'll try that.
